### ISSUE TITLE
sam unable to connect endpoint url

### ISSUE THREAD

**Issue by mrigankt** [awslabs/aws-sam-local#102]
we created a lambda function such that when it is executed it makes a entry in dynamob db local. and for it need to connect to the endpoint url 127.0.0.1:8000, but it is unable to connect with following error:-
```
Error JSON: {
  "message": "connect ECONNREFUSED 127.0.0.1:8000",
  "code": "NetworkingError",
  "errno": "ECONNREFUSED",
  "syscall": "connect",
  "address": "127.0.0.1",
  "port": 8000,
  "region": "us-west-2",
  "hostname": "localhost",
  "retryable": true,
  "time": "2017-08-30T07:56:18.509Z"
}
```

For testing purpose  i used  node and lambda-local to execute my lambda function and it is working perfectly fine. below is my lambda function and template 

index.js

> 'use strict';
> console.log('starting function');
> 
> 
> var AWS = require("aws-sdk");
> 
> AWS.config.update({
>   region: "us-east-1",
>   endpoint: "http://localhost:8000"
> });
> 
> var dynamodb = new AWS.DynamoDB.DocumentClient();
> 
> 
> 
> exports.handle = function(e, ctx, callback) {
> 
>        var params = {
>                 Item: {
>                         date: Date.now(),
>                         message: "Hello how are you"
>                 },
> 
>                TableName: 'guestbook'
>         };
> 
>        dynamodb.put(params, function(err, data){
>                 if(err){
>                         callback(err, null);
>                         }else{
>                         callback(null, data);
>                         }
>                 });
> 
>        }
> 

template :- 

> AWSTemplateFormatVersion : '2010-09-09'
> Transform: AWS::Serverless-2016-10-31
> 
> Description: dynamodb
> Resources:
>   dynamodb:
>     Type: AWS::Serverless::Function
>     Properties:
>       Timeout: 60
>       Runtime: nodejs6.10
>       Handler: index.handle
>       Policies: AmazonDynamoDBFullAccess

**Comment by splityo** [comment]
I had the same problem and the only explanation I found is is that the docker container is NOT running the DynamoDB Local process INSIDE the container (localhost). My solution was to use the IP address assigned to the network interface and update the endpoint to that address, with this workaround I was able to connect to DynamoDB Local. Ex:

AWS.config.update({
endpoint: "http://NETWORK_INTERFACE_ADDRESS:8000"
});

Please let me know if this works for you.

Regards

**Comment by PaulMaddox** [comment]
That's correct @splityo.
Closing this issue - @mrigankt, feel free to re-open if the above doesn't help.

**Comment by mrigankt** [comment]
Thank you 😀

**Comment by TatchNicolas** [comment]
Is it possible to check if the alias on loopback is set correctly by calling

 `aws dynamodb --endpoint-url http://172.16.123.1 <any operation on DynamoDBLocal>` 

In other words, should it be possible to say endpoint URL is reachable from SAM Local if the command above is executable on the host machine which runs `sam local start-api` ?

**Comment by heitorlessa** [comment]
If that helps, I actually run DynamoDB Local as a Docker Container in the
same Network as my Lambda functions spun up by SAM Local. From there, i
simply name my DynamoDB Local container as "dynamo" and inside my code i
simply connect to "dynamo".

*Some info*: https://github.com/heitorlessa/sam-local-python-hot-reloading
*Docker network*: docker network create sam-demo
*DynamoDB Local Container*: docker run -d -v "$PWD":/dynamodb_local_db -p
8000:8000 --network sam-demo --name dynamodb cnadiminti/dynamodb-local
*SAM Local using the same network*: sam local start-api --docker-network
sam-demo

**Comment by TatchNicolas** [comment]
@heitorlessa Thanks, your way of setting looks nice, I'll check that out.
For the question I posted I sorted out by myself it turns out to be true, it works fine, I just did not set the alias correctly.

**Comment by jfernachewy** [comment]
Just set your redis host as 'host.docker.internal'

**Comment by fjbarrientosg** [comment]
> I had the same problem and the only explanation I found is that the docker container is NOT running the DynamoDB Local process INSIDE the container (localhost) then is unable to reach out. My solution was to assign an IP alias to loopback to reach out using a different address to 127.0.0.1, and update the endpoint to that address, with this workaround I was able to connect to DynamoDB Local without network connection. Example:
> 
> At the command line run
> `ifconfig lo0 alias 172.16.123.1`
> 
> Update your endpoint to use the alias and you are done
> 
> ```
> AWS.config.update({
>     endpoint: "http://172.16.123.1:8000"
> });
> ```
> 
> To remove the alias
> `ifconfig lo0 -alias 172.16.123.1`
> 
> Do not forget to add "sudo" where applicable (Mac, Ubuntu, etc)
> 
> Please let me know if this works for you.
> 
> Regards

I'm not sure but I was facing same issue but with AWS SAM, trying to invoke a lambda function which invokes another lambda function inside it. So I could execute the parent lambda function from my host or localhost without problem, but my internal lambda function got failed because it could not connect to my host localhost. So I needed to configurate a special endpoint to my internal function with **host.docker.internal** (https://stackoverflow.com/questions/24319662/from-inside-of-a-docker-container-how-do-i-connect-to-the-localhost-of-the-mach). Hope it helps (srry my english lol)

Basically my host lambda configuration is looks like this:

```
LAMBDA_CONFIG: {
        apiVersion: '2015-03-31',
        region: 'localhost',
        endpoint: 'http://127.0.0.1:3001',
        sslEnabled: false
    }
```

And my internal (container):

```
LAMBDA_CONTAINER_CONFIG: {
        apiVersion: '2015-03-31',
        region: 'localhost',
        endpoint: 'http://host.docker.internal:3001',
        sslEnabled: false
    }
```

**Comment by NisargChokshi45** [comment]
> we created a lambda function such that when it is executed it makes a entry in dynamob db local. and for it need to connect to the endpoint url 127.0.0.1:8000, but it is unable to connect with following error:-
> 
> ```
> Error JSON: {
>   "message": "connect ECONNREFUSED 127.0.0.1:8000",
>   "code": "NetworkingError",
>   "errno": "ECONNREFUSED",
>   "syscall": "connect",
>   "address": "127.0.0.1",
>   "port": 8000,
>   "region": "us-west-2",
>   "hostname": "localhost",
>   "retryable": true,
>   "time": "2017-08-30T07:56:18.509Z"
> }
> ```
> 
> For testing purpose i used node and lambda-local to execute my lambda function and it is working perfectly fine. below is my lambda function and template
> 
> index.js
> 
> > 'use strict';
> > console.log('starting function');
> > var AWS = require("aws-sdk");
> > AWS.config.update({
> > region: "us-east-1",
> > endpoint: "http://localhost:8000"
> > });
> > var dynamodb = new AWS.DynamoDB.DocumentClient();
> > exports.handle = function(e, ctx, callback) {
> > ```
> >    var params = {
> >             Item: {
> >                     date: Date.now(),
> >                     message: "Hello how are you"
> >             },
> > 
> >            TableName: 'guestbook'
> >     };
> > 
> >    dynamodb.put(params, function(err, data){
> >             if(err){
> >                     callback(err, null);
> >                     }else{
> >                     callback(null, data);
> >                     }
> >             });
> > 
> >    }
> > ```
> 
> template :-
> 
> > AWSTemplateFormatVersion : '2010-09-09'
> > Transform: AWS::Serverless-2016-10-31
> > Description: dynamodb
> > Resources:
> > dynamodb:
> > Type: AWS::Serverless::Function
> > Properties:
> > Timeout: 60
> > Runtime: nodejs6.10
> > Handler: index.handle
> > Policies: AmazonDynamoDBFullAccess

I came across the Same Error. I am using Ubuntu 20.04 & I have changed my local AWS configuration endpoint to the IP address of the Docker service

You can check the IP using the following command:
> ifconfig -a 

Find the IP of the Docker service & Replace with your AWS endpoint. This solves my issue.
